### PR TITLE
feat: implement CosmWasm context for light clients + contract for Tendermint clients

### LINKED ISSUE #1164: Provide library to facilitate CosmWasm contract implementation for light clients built using ibc-rs
## Summary
The goal is to integrate a library into ibc-rs that includes essential common definitions and utilities, ready-made to facilitate the implementation of CosmWasm contracts for light clients built using ibc-rs.

This initiative is inspired by our experience in implementing the [CosmWasm contract](https://github.com/informalsystems/sovereign-ibc/tree/main/clients/sov-celestia-cw) for Sovereign SDK light client.

This library, when added to ibc-rs, will provide users with a comprehensive `Context` object. This context will have all the necessary methods and client-level traits already implemented and intended to be compatible with any light clients developed using ibc-rs.

```[tasklist]
### Prerequisite
- [ ] https://github.com/cosmos/ibc-rs/issues/1163
```

### LINKED ISSUE #1165: Build CosmWasm contract for ICS-07 Tendermint light client
## Overview
With the shared library described [here](https://github.com/cosmos/ibc-rs/issues/1164), we can create a CosmWasm contract for a light client built using ibc-rs. This means we can include a basic Rust crate for the ICS-07 Tendermint client. In this crate, we'll utilize the API provided by the mentioned library. This includes defining our desired client and consensus types, setting up a Tendermint-specific context, and then using that object in the contract's entry point.

```[tasklist]
### Prerequisite
- [ ] https://github.com/cosmos/ibc-rs/issues/1164
```

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Comes after: #1163

Closes: #1164
Closes: #1165

Subsequent PR on sovereign-ibc: https://github.com/informalsystems/sovereign-ibc/pull/142

## Description
This PR adds the following crates:
- `ibc-client-cw` A library providing ready-made `Context` object to facilitate CosmWasm integration
- `ibc-client-tendermint-cw`: Implemtens CosmWasm contract for ICS-07 Tendermint light clients

Plus, wrote unit tests for Tendermint CosmWasm client creation / update / recovery.

## TODO in follow-up PRs
- Include CI job to check the `make build-tendermint-cw`
- Add Nix flake to build `ibc_client_tendermint_cw.wasm` file
- Write macro to auto-generate the entrypoints and handle some required conversions. 

______

### PR author checklist:
- [X] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [X] Linked to GitHub issue.
- [X] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -29,5 +29,7 @@ mc.log
 # Specifically, we want the ci/no-std-check/Cargo.lock to be committed
 # to ensure a reproducible CI build for that crate.
 # This will make it easier to spot bugs such as [this one](https://github.com/cosmos/ibc-rs/pull/560)
-# caused by a dependency updated to a faulty version. 
+# caused by a dependency updated to a faulty version.
 /Cargo.lock
+
+/cw-contracts
```
Comment (rnbguy) [inline .gitignore]:
```suggestion
# Ignore CosmWasm artifacts
/cw-contracts

```

File: .changelog/unreleased/features/1165-implement-cosmwasm-contract-for-tendermint-clients.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+- [ibc-client-tendermint] Implement CosmWasm contract for ICS-07 Tendermint
+  light clients available as `ibc-client-tendermint-cw` crate.
+- ([\#1165](https://github.com/cosmos/ibc-rs/issues/1165))
```
Comment (rnbguy) [inline .changelog/unreleased/features/1165-implement-cosmwasm-contract-for-tendermint-clients.md]:
```suggestion
- [ibc-client-tendermint] Introduce `ibc-client-tendermint-cw` crate
  implementing ICS-07 Tendermint light client as ICS-008 CosmWasm light client.
  ([\#1165](https://github.com/cosmos/ibc-rs/issues/1165))
```

File: .changelog/unreleased/features/1164-implement-cosmwasm-integrator-library.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+- [ibc-clients] Implement CosmWasm integrator library available as
+`ibc-client-cw` crate to facilitate CosmWasm contract creation for light clients
+built using `ibc-rs` ([\#1164](https://github.com/cosmos/ibc-rs/issues/1164))
```
Comment (rnbguy) [inline .changelog/unreleased/features/1164-implement-cosmwasm-integrator-library.md]:
```suggestion
- [ibc-clients] Introduce `ibc-client-cw` crate to implement ICS-008 CosmWasm
  light clients using `ibc-rs`.
  ([\#1164](https://github.com/cosmos/ibc-rs/issues/1164))
```

File: Cargo.toml [hunk 4]
```diff
@@ -62,22 +65,28 @@ serde_json          = { package = "serde-json-wasm", version = "1.0.1", default-
...
-ibc-clients         = { version = "0.51.0", path = "./ibc-clients", default-features = false }
-ibc-apps            = { version = "0.51.0", path = "./ibc-apps", default-features = false }
-ibc-primitives      = { version = "0.51.0", path = "./ibc-primitives", default-features = false }
-ibc-derive          = { version = "0.6.1", path = "./ibc-derive" }
+ibc                         = { version = "0.51.0", path = "./ibc", default-features = false }
+ibc-core                    = { version = "0.51.0", path = "./ibc-core", default-features = false }
+ibc-clients                 = { version = "0.51.0", path = "./ibc-clients", default-features = false }
+ibc-apps                    = { version = "0.51.0", path = "./ibc-apps", default-features = false }
+ibc-primitives              = { version = "0.51.0", path = "./ibc-primitives", default-features = false }
+ibc-testkit                 = { version = "0.51.0", path = "./ibc-testkit", default-features = false }
```
Comment (rnbguy) [inline Cargo.toml:73]: Is `ibc-testkit` is necessary here? I don't think that any other sub-crates depends on it.

File: makefile [hunk 5]
```diff
@@ -40,3 +40,9 @@ check-release: ## Check that the release build compiles.
 
 release: ## Perform an actual release and publishes to crates.io.
 	cargo release --workspace --no-push --no-tag --exclude ibc-derive --allow-branch HEAD --execute
+
+build-tendermint-cw: ## Build the WASM file for the ICS-07 Tendermint light client.
+	@echo "Building the WASM file for the ICS-07 Tendermint light client"
+	@RUSTFLAGS='-C link-arg=-s' cargo build -p ibc-client-tendermint-cw --target wasm32-unknown-unknown --release --lib --locked
+	@mkdir -p cw-contracts
+	@cp target/wasm32-unknown-unknown/release/ibc_client_tendermint_cw.wasm cw-contracts/
```
Comment (rnbguy) [inline makefile]: Do we need to [suppress the echo of this command](https://www.gnu.org/software/make/manual/make.html#Recipe-Echoing)? If it is, makes sense to make this consistent over this `Makefile`

File: ibc-clients/README.md [hunk 6]
```diff
@@ -23,13 +23,25 @@ IBC light clients:
...
+### CosmWasm Integration
+
+- [ibc-client-cw](./cw-context): Types and Utilities for CosmWasm Integration
+  - To utilize the CosmWasm contracts developed with this library, hosting
+    environments must support the CosmWasm module and be using the version of
+    `ibc-go` that supports the `08-wasm` proxy light client.
+
+  - **DISCLAIMER: This library is currently in development and should not be
+    deployed for production use. Users are advised to exercise caution and test
+    thoroughly in non-production environments.**
```
Comment (rnbguy) [inline ibc-clients/README.md]: You can use [native Github markdown syntax](https://github.com/orgs/community/discussions/16925) for the disclaimer.

File: ibc-clients/ics07-tendermint/cw-contract/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ibc-core   			  = { workspace = true }
+ibc-client-cw         = { workspace = true }
+ibc-client-tendermint = { workspace = true }
+
+# cosmwasm dependencies
+cosmwasm-std 	      = { workspace = true }
+
+[dev-dependencies]
+cosmwasm-vm           = { workspace = true }
+hex				      = "0.4.2"
```
Comment (rnbguy) [inline ibc-clients/ics07-tendermint/cw-contract/Cargo.toml]:
```suggestion
hex				      = { version = "0.4.2" }
```

File: ibc-clients/ics07-tendermint/cw-contract/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,42 @@
+[package]
```
Comment (rnbguy) [inline ibc-clients/ics07-tendermint/cw-contract/Cargo.toml:1]: This file contains tabs (`\t`) - make sure to convert them to spaces.

File: ibc-clients/ics07-tendermint/cw-contract/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+authors 	 = { workspace = true }
+edition 	 = { workspace = true }
+license 	 = { workspace = true }
+repository   = { workspace = true }
+rust-version = { workspace = true }
+version 	 = { workspace = true }
+keywords     = ["ibc", "light-client", "ics07", "tendermint", "CosmWasm"]
+readme       = "./../../README.md"
+description  = """
+    The CosmWasm contract implementation of the ICS-07 Tendermint light client.
```
Comment (rnbguy) [inline ibc-clients/ics07-tendermint/cw-contract/Cargo.toml]:
```suggestion
    The ICS-008 CosmWasm contract implementation of the ICS-07 Tendermint light client.
```

File: ibc-clients/ics07-tendermint/cw-contract/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,42 @@
+[package]
+name 		 = "ibc-client-tendermint-cw"
+authors 	 = { workspace = true }
+edition 	 = { workspace = true }
+license 	 = { workspace = true }
+repository   = { workspace = true }
+rust-version = { workspace = true }
+version 	 = { workspace = true }
+keywords     = ["ibc", "light-client", "ics07", "tendermint", "CosmWasm"]
```
Comment (rnbguy) [inline ibc-clients/ics07-tendermint/cw-contract/Cargo.toml:9]: add `ics008` in the keywords.

File: ibc-clients/cw-context/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,40 @@
+[package]
+name 		 = "ibc-client-cw"
+authors 	 = { workspace = true }
+edition 	 = { workspace = true }
+license 	 = { workspace = true }
+repository   = { workspace = true }
+rust-version = { workspace = true }
+version 	 = { workspace = true }
+keywords     = ["ibc", "light-client", "CosmWasm"]
```
Comment (rnbguy) [inline ibc-clients/cw-context/Cargo.toml]: add `ics008` in the keywords.

File: ibc-clients/cw-context/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+[features]
+default = ["std"]
+std = [
+	"prost/std",
+	"serde/std",
+	"ibc-core/std",
+	"ibc-client-wasm-types/std",
+]
+
```
Comment (rnbguy) [inline ibc-clients/cw-context/Cargo.toml]:
```suggestion
```

File: ibc-clients/cw-context/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,40 @@
+[package]
```
Comment (rnbguy) [inline ibc-clients/cw-context/Cargo.toml:1]: replace tabs(`\t`) using spaces in this file.

File: ibc-clients/cw-context/src/api.rs [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+use ibc_core::primitives::proto::Any;
+
+use crate::context::Context;
+
+/// Enables users to integrate their implemented light client by introducing
+/// their client state and consensus state types into the generic [`Context`]
+/// object.
+pub trait ClientType<'a>: Sized {
+    type ClientState: ClientStateExecution<Context<'a, Self>> + Clone;
+    type ConsensusState: ConsensusStateTrait + Into<Any> + TryFrom<Any, Error = ClientError>;
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/api.rs:13]: Can we use the `Convertible` trait here? If it is available via dependencies - otherwise, it's ok.

Comment (rnbguy) [reply ibc-clients/cw-context/src/api.rs:13]: Maybe we can move `Convertible` trait to `ibc_primitives`.

File: ibc-clients/cw-context/src/context/client_ctx.rs [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    fn client_update_meta(
+        &self,
+        _client_id: &ClientId,
+        height: &Height,
+    ) -> Result<(Timestamp, Height), ContextError> {
+        let time_key = self.client_update_time_key(height);
+
+        let time_vec = self.retrieve(time_key)?;
+
+        let time = u64::from_be_bytes(time_vec.try_into().expect("invalid timestamp"));
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/context/client_ctx.rs]: does it make sense to convert the `.expect` to a `Err`?

File: ibc-clients/cw-context/src/context/client_ctx.rs [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+
+        let timestamp =
+            Timestamp::from_nanoseconds(time).map_err(ClientError::InvalidPacketTimestamp)?;
+
+        let height_key = self.client_update_height_key(height);
+
+        let revision_height_vec = self.retrieve(height_key)?;
+
+        let revision_height =
+            u64::from_be_bytes(revision_height_vec.try_into().expect("invalid height"));
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/context/client_ctx.rs]: calling `.expect` here. maybe propagate the `Err`.

File: ibc-clients/cw-context/src/context/custom_ctx.rs [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+
+use super::Context;
+use crate::api::ClientType;
+use crate::types::HeightTravel;
+
+impl<'a, C: ClientType<'a>> ExtClientValidationContext for Context<'a, C> {
+    fn host_timestamp(&self) -> Result<Timestamp, ContextError> {
+        let time = self.env().block.time;
+
+        let host_timestamp = Timestamp::from_nanoseconds(time.nanos()).expect("invalid timestamp");
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/context/custom_ctx.rs]: calling `.expect`.

File: ibc-clients/cw-context/src/types/msgs.rs [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+
+impl TryFrom<VerifyNonMembershipMsgRaw> for VerifyNonMembershipMsg {
+    type Error = ContractError;
+
+    fn try_from(mut raw: VerifyNonMembershipMsgRaw) -> Result<Self, Self::Error> {
+        let proof = CommitmentProofBytes::try_from(raw.proof)?;
+        let prefix = raw.path.key_path.remove(0).into_bytes();
+        let path_str = raw.path.key_path.join("");
+        let path = Path::from_str(&path_str)?;
+        let height = raw.height.try_into().expect("invalid height");
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/types/msgs.rs]: calling `.expect`.

File: ibc-clients/cw-context/src/context/client_ctx.rs [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+
+        let time_vec: [u8; 8] = host_timestamp.nanoseconds().to_be_bytes();
+
+        self.insert(prefixed_time_key, time_vec);
+
+        let height_key = self.client_update_height_key(&height);
+
+        let prefixed_height_key = self.prefixed_key(height_key);
+
+        let revision_height_vec: [u8; 8] = host_height.revision_height().to_be_bytes();
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/context/client_ctx.rs]: Do we need this `: [u8; 8]`? [`u64::to_be_bytes` returns `[u8; 8]`](https://doc.rust-lang.org/std/primitive.u64.html#method.to_be_bytes).

There are a few others.

File: ibc-clients/cw-context/src/context/custom_ctx.rs [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+            Some(prev_height) => {
+                let cons_state_path = ClientConsensusStatePath::new(
+                    client_id.clone(),
+                    prev_height.revision_number(),
+                    prev_height.revision_height(),
+                );
+                self.consensus_state(&cons_state_path).map(Some)
+            }
+            None => Ok(None),
+        }
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/context/custom_ctx.rs:75]: I think, you can use
```rs
prev_height.map(|prev_height| let cons_state_path = ClientConsensusStatePath::new(
                    client_id.clone(),
                    prev_height.revision_number(),
                    prev_height.revision_height(),
                );
                self.consensus_state(&cons_state_path)).transpose()
```

File: ibc-clients/cw-context/src/context/mod.rs [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+pub trait StorageMut: StorageRef {
+    fn storage_mut(&mut self) -> &mut dyn Storage;
+}
+
+impl<'a, C: ClientType<'a>> StorageMut for Context<'a, C> {
+    fn storage_mut(&mut self) -> &mut dyn Storage {
+        match self.deps_mut {
+            Some(ref mut deps) => deps.storage,
+            None => panic!("deps_mut should be available"),
+        }
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/context/mod.rs:290]: You should be able to do following:

```rs
self.deps_mut.as_ref_mut().map(|deps| deps.storage).expect("deps_mut should be available");
```

File: ibc-clients/cw-context/src/context/mod.rs [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+impl<'a, C: ClientType<'a>> StorageRef for Context<'a, C> {
+    fn storage_ref(&self) -> &dyn Storage {
+        match self.deps {
+            Some(ref deps) => deps.storage,
+            None => match self.deps_mut {
+                Some(ref deps) => deps.storage,
+                None => panic!("Either deps or deps_mut should be available"),
+            },
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/context/mod.rs:276]:
```rs
self.deps.as_ref().map(|deps| deps.storage)
.or_else(|_| self.deps_mut.as_ref().map(|deps| deps.storage))
.expect("Either deps or deps_mut should be available")
```

File: ibc-clients/cw-context/src/handlers.rs [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+                    substitute_client_state.into(),
+                )?;
+
+                ContractResult::success()
+            }
+        };
+        Ok(to_json_binary(&result)?)
+    }
+
+    /// Queries the client with the given `QueryMsg` message.
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/handlers.rs]:
```suggestion
    /// Queries the client with the given [`QueryMsg`] message.
```

File: ibc-clients/cw-context/src/handlers.rs [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+        let any_consensus_state = Any::decode(&mut msg.consensus_state.as_slice())?;
+
+        self.set_checksum(msg.checksum);
+
+        client_state.initialise(self, &self.client_id(), any_consensus_state)?;
+
+        Ok(to_json_binary(&ContractResult::success())?)
+    }
+
+    /// Executes the sudo message.
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/handlers.rs]:
```suggestion
    /// Executes the given [`SudoMsg`].
```

File: ibc-clients/cw-context/src/handlers.rs [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+use crate::context::Context;
+use crate::types::{
+    CheckForMisbehaviourMsg, ContractError, ContractResult, ExportMetadataMsg, InstantiateMsg,
+    QueryMsg, QueryResponse, StatusMsg, SudoMsg, UpdateStateMsg, UpdateStateOnMisbehaviourMsg,
+    VerifyClientMessageMsg, VerifyMembershipMsg, VerifyNonMembershipMsg,
+    VerifyUpgradeAndUpdateStateMsg,
+};
+
+impl<'a, C: ClientType<'a>> Context<'a, C> {
+    /// Instantiates a new client with the given `InstantiateMsg` message.
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/handlers.rs]:
```suggestion
    /// Instantiates a new client with the given [`InstantiateMsg`] message.
```

File: ibc-clients/cw-context/src/types/error.rs [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+use ibc_core::client::types::error::ClientError;
+use ibc_core::commitment_types::error::CommitmentError;
+use ibc_core::handler::types::error::ContextError;
+use ibc_core::host::types::error::IdentifierError;
+use ibc_core::host::types::path::PathError;
+use prost::DecodeError;
+
+#[derive(From, Display, Debug)]
+pub enum ContractError {
+    #[display(fmt = "CosmWasm standard error: {_0}")]
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/types/error.rs:12]: interesting ! I didn't know `derive_more` has `Display` too.

we were using `displaydoc` in other sub-crates.

File: ibc-clients/cw-context/src/types/helper.rs [hunk 27]
```diff
@@ -0,0 +1,30 @@
+use ibc_client_wasm_types::{SUBJECT_PREFIX, SUBSTITUTE_PREFIX};
+
+/// The MigrationPrefix enumerates the prefix type used during migration mode.
+/// The migration mode is activated when there is an incoming
+/// `MigrateClientStore` message. It specifies the prefix key for either the
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/types/helper.rs]: We may link `MigrateClientStore` in docs.

File: ibc-clients/cw-context/src/types/msgs.rs [hunk 28]
```diff
@@ -0,0 +1,296 @@
...
+use ibc_core::client::types::proto::v1::Height as RawHeight;
+use ibc_core::client::types::Height;
+use ibc_core::commitment_types::commitment::{CommitmentPrefix, CommitmentProofBytes};
+use ibc_core::host::types::path::Path;
+use ibc_core::primitives::proto::Any;
+use prost::Message;
+
+use super::error::ContractError;
+
+pub type Bytes = Vec<u8>;
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/types/msgs.rs]: Does it make sense to put it in some `utils` module? Or even move to `ibc_primitives` prelude?

File: ibc-clients/cw-context/src/types/response.rs [hunk 29]
```diff
@@ -0,0 +1,71 @@
+//! Contains the response types for the CosmWasm contract.
+use cosmwasm_schema::cw_serde;
+use ibc_core::client::types::Height;
+
+#[cw_serde]
+pub struct GenesisMetadata {
+    pub key: Vec<u8>,
+    pub value: Vec<u8>,
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/types/response.rs]: use `Bytes` from other module.

File: ibc-clients/cw-context/src/utils/codec.rs [hunk 30]
```diff
@@ -0,0 +1,27 @@
+use ibc_core::client::types::error::ClientError;
+use ibc_core::primitives::proto::Any;
+use prost::Message;
+
+/// AnyCodec is a convenient trait that provides a generic way to encode and
+/// decode domain types through the `Any` type.
+pub trait AnyCodec {
+    fn decode_thru_any<C>(data: Vec<u8>) -> Result<C, ClientError>
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/utils/codec.rs]:
```suggestion
    fn decode_any_vec<C>(data: Vec<u8>) -> Result<C, ClientError>
```

File: ibc-clients/cw-context/src/utils/codec.rs [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+        C: TryFrom<Any, Error = ClientError>,
+    {
+        let raw = Any::decode(&mut data.as_slice()).map_err(|e| ClientError::Other {
+            description: e.to_string(),
+        })?;
+
+        C::try_from(raw)
+    }
+
+    fn encode_thru_any<C>(value: C) -> Vec<u8>
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/utils/codec.rs]:
```suggestion
    fn encode_thru_any<C>(value: C) -> Vec<u8>
```
```suggestion
    fn encode_to_any_vec<C>(value: C) -> Vec<u8>
```

File: ibc-clients/cw-context/src/utils/codec.rs [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+    where
+        C: Into<Any>,
+    {
+        value.into().encode_to_vec()
+    }
+}
+
+impl<T> AnyCodec for T where T: TryFrom<Any, Error = ClientError> + Into<Any> {}
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/utils/codec.rs:27]: I think that I commented earlier about this being similar to `Convertible`. But this does the conversion via `protobuf::Any` type.

File: ibc-clients/cw-context/src/utils/mod.rs [hunk 33]
```diff
@@ -0,0 +1,22 @@
...
+
+pub use codec::*;
+use ibc_core::client::types::error::ClientError;
+use ibc_core::client::types::{Height, HeightError};
+
+/// Decodes a `Height` from a UTF-8 encoded byte array.
+pub fn parse_height(encoded_height: Vec<u8>) -> Result<Option<Height>, ClientError> {
+    let height_str = match alloc::str::from_utf8(encoded_height.as_slice()) {
+        Ok(s) => s,
+        Err(_) => return Ok(None),
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/utils/mod.rs:14]: Can you please add the comment here too?

File: ibc-clients/ics07-tendermint/cw-contract/src/tests/helper.rs [hunk 34]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+pub fn dummy_checksum() -> Vec<u8> {
+    hex::decode("2469f43c3ca20d476442bd3d98cbd97a180776ab37332aa7b02cae5a620acfc6")
+        .expect("Never fails")
+}
+
+pub fn dummy_sov_consensus_state(timestamp: Timestamp) -> ConsensusState {
+    ConsensusState::new(
+        vec![0].into(),
```
Comment (rnbguy) [inline ibc-clients/ics07-tendermint/cw-contract/src/tests/helper.rs:20]: This should work too.

```suggestion
        &[0].into(),
```

**Review by rnbguy — CHANGES_REQUESTED** [review]
Thanks for this cool PR :rocket:

Added review comments!

File: Cargo.toml (hunk 4 above)
Comment (Farhad-Shabani) [reply Cargo.toml:73]: It's been added under the [dev-dependency] of the `ibc-client-tendermint-cw`.

File: ibc-clients/cw-context/src/api.rs (hunk 14 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/api.rs:13]: Actually, can't use that. As here we are interested in the conversions opposite to those `Convertible` offers. 

> Maybe we can move Convertible trait to ibc_primitives.

Even though `Convertible` is quite versatile, IMO, for now it's best to keep it within the ICS-02 considering its specific use case.

File: .gitignore (hunk 1 above)
Comment (Farhad-Shabani) [reply .gitignore]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: .changelog/unreleased/features/1165-implement-cosmwasm-contract-for-tendermint-clients.md (hunk 2 above)
Comment (Farhad-Shabani) [reply .changelog/unreleased/features/1165-implement-cosmwasm-contract-for-tendermint-clients.md]: Note that ICS-008 is a separate type of light client that acts as a proxy redirecting client-relevant messages to the CosmWasm contracts. By the way, applied a minor update to the log.

File: .changelog/unreleased/features/1164-implement-cosmwasm-integrator-library.md (hunk 3 above)
Comment (Farhad-Shabani) [reply .changelog/unreleased/features/1164-implement-cosmwasm-integrator-library.md]: https://github.com/cosmos/ibc-rs/pull/1167#discussion_r1569172602

File: ibc-clients/README.md (hunk 6 above)
Comment (Farhad-Shabani) [reply ibc-clients/README.md]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/ics07-tendermint/cw-contract/Cargo.toml (hunk 7 above)
Comment (Farhad-Shabani) [reply ibc-clients/ics07-tendermint/cw-contract/Cargo.toml]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/ics07-tendermint/cw-contract/Cargo.toml (hunk 8 above)
Comment (Farhad-Shabani) [reply ibc-clients/ics07-tendermint/cw-contract/Cargo.toml:1]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/ics07-tendermint/cw-contract/Cargo.toml (hunk 9 above)
Comment (Farhad-Shabani) [reply ibc-clients/ics07-tendermint/cw-contract/Cargo.toml]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/Cargo.toml (hunk 11 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/Cargo.toml]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/Cargo.toml (hunk 13 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/Cargo.toml:1]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/context/client_ctx.rs (hunk 15 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/context/client_ctx.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/context/custom_ctx.rs (hunk 17 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/context/custom_ctx.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/context/client_ctx.rs (hunk 16 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/context/client_ctx.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/types/msgs.rs (hunk 18 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/types/msgs.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/context/custom_ctx.rs (hunk 20 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/context/custom_ctx.rs:75]: Can we keep this as is? I am choosing more readable, quicker to grasp over the efficiency here.

File: ibc-clients/cw-context/src/context/mod.rs (hunk 21 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/context/mod.rs:290]: https://github.com/cosmos/ibc-rs/pull/1167#discussion_r1569178327

File: ibc-clients/cw-context/src/context/mod.rs (hunk 22 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/context/mod.rs:276]: https://github.com/cosmos/ibc-rs/pull/1167#discussion_r1569178327

File: ibc-clients/cw-context/src/types/helper.rs (hunk 27 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/types/helper.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/types/msgs.rs (hunk 28 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/types/msgs.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/types/response.rs (hunk 29 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/types/response.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/utils/codec.rs (hunk 30 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/utils/codec.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/utils/codec.rs (hunk 31 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/utils/codec.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/ics07-tendermint/cw-contract/src/tests/helper.rs (hunk 34 above)
Comment (Farhad-Shabani) [reply ibc-clients/ics07-tendermint/cw-contract/src/tests/helper.rs:20]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/utils/mod.rs (hunk 33 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/utils/mod.rs:14]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: ibc-clients/cw-context/src/context/client_ctx.rs (hunk 19 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/context/client_ctx.rs]: [6de21ed](https://github.com/cosmos/ibc-rs/pull/1167/commits/6de21edb62137b9d74daf1bcacf3b626e45013cc)

File: makefile (hunk 5 above)
Comment (Farhad-Shabani) [reply makefile]: [80e1de4](https://github.com/cosmos/ibc-rs/pull/1167/commits/80e1de4c4327d306491d17b760665bbe0ffd96d9)

File: ibc-clients/ics07-tendermint/cw-contract/Cargo.toml (hunk 10 above)
Comment (Farhad-Shabani) [reply ibc-clients/ics07-tendermint/cw-contract/Cargo.toml:9]: We can just have up to five keywords :D

**Review by rnbguy — APPROVED** [review]
LGTM :rocket:

I added a few minor commits. Feel free to adjust them.

**Pull request merged**